src/metabase/models/permissions.clj

```diff
@@ -853,6 +927,41 @@
   (delete-related-permissions! group-or-id (apply (partial feature-perms-path :download :full) path-components))
   (delete-related-permissions! group-or-id (apply (partial feature-perms-path :download :limited) path-components)))
 
+(let [delete (fn delete [s to-delete] (str/replace s to-delete ""))
+      data-query-split (fn data-query-split [path]
+                         [(str "/data" path) (str "/query" path)])]
+  (def ^:private data-kind->rewrite-fn
+    "lookup table to generate v2 query + data permission from a v1 data permission."
+    {:dk/db                                 data-query-split
+     :dk/db-native                          (fn [path] (data-query-split (delete path "native/")))
+     :dk/db-schema                          (fn [path] [(str "/data" (delete path "schema/"))
+                                                        (str "/query" path)])
+     :dk/db-schema-name                     data-query-split
+     :dk/db-schema-name-and-table           data-query-split
+     :dk/db-schema-name-table-and-read      (constantly [])
+     :dk/db-schema-name-table-and-query     (fn [path] (data-query-split (delete path "query/")))
+     :dk/db-schema-name-table-and-segmented (fn [path] (data-query-split (delete path "query/segmented/")))}))
+
+
+
+(mu/defn ^:private ->v2-path :- [:vector [:re path-regex-v2]]
+  [path :- [:or [:re path-regex-v1] [:re path-regex-v2]]]
+  ;; See: https://www.notion.so/metabase/Permissions-Refactor-Design-Doc-18ff5e6be32f4a52b9422bd7f4237ca7#5603afe084a7435ca7dc928fc94d4bda
+  (let [kind (classify-path path)]
+    (cond
+      (= kind :data) (let [data-permission-kind (classify-data-path path)
+                           rewrite-fn (data-kind->rewrite-fn data-permission-kind)]
+                       (rewrite-fn path))
+      (= kind :admin) ["/"]
+      (= kind :block) []
+
+      ;; explicitly, move is idempotent mapcatting move over a sequence of v1 paths multiple times will result in the same value.
+      ;; so these are no-ops:
+      (= kind :data-v2) [path]
+      (= kind :query-v2) [path]
```

Comment:
Should this comment explain why these cases are explicitly listed?

Reply:
It should. I'll make it more clear.